File: example/cpp/src/product/g2_force.cpp
#include <argparse/argparse.hpp>
#include <boost/asio.hpp>
#include <boost/json.hpp>
#include <cmath>
#include <csignal>
#include <iostream>
#include <thread>
#include <utility>

#include "airbot_hardware/executors/executor.hpp"
#include "airbot_hardware/handlers/arm.hpp"

using namespace std::chrono_literals;
using namespace airbot::hardware;

static bool running = true;  // 全局标志，用于控制线程和循环

void signal_handler(int signal) {
  if (signal == SIGINT) {
    std::cout << "\nCaught CTRL+C, exiting..." << std::endl;
    running = false;
  }
}

boost::json::object commandToJson(const std::string& command_type, const std::vector<double>& positions,
                                  const std::vector<double>& velocity, const std::vector<double>& effort) {
  boost::json::object json_obj;

  boost::json::array pos_array;
  boost::json::array vel_array;
  boost::json::array eff_array;
  for (const auto& pos : positions) pos_array.push_back(pos);
  for (const auto& vel : velocity) vel_array.push_back(vel);
  for (const auto& eff : effort) eff_array.push_back(eff);

  json_obj[command_type + "/position"] = pos_array;
  json_obj[command_type + "/velocity"] = vel_array;
  json_obj[command_type + "/effort"] = eff_array;
  json_obj["timestamp"] = static_cast<double>(std::chrono::duration_cast<std::chrono::microseconds>(
                                                  std::chrono::steady_clock::now().time_since_epoch())
                                                  .count()) /
                          1e6;

  return json_obj;
}

int main(int argc, char* argv[]) {
  std::signal(SIGINT, signal_handler);

  argparse::ArgumentParser program("g2_pos_ctrl");

  program.add_argument("-i").help("can port").default_value(std::string("can0"));
  program.add_argument("-r", "--remote-ip").help("Plotjuggler IP address").default_value(std::string("127.0.0.1"));
  program.add_argument("-p", "--port").help("UDP port for Plotjuggler").scan<'i', int>().default_value(9870);
  program.add_argument("-f").help("eef eff").scan<'g', double>().default_value(5.0);

  try {
    program.parse_args(argc, argv);
  } catch (const std::runtime_error& err) {
    std::cerr << err.what() << std::endl;
    std::cerr << program;
    return 1;
  }

  auto executor = airbot::hardware::AsioExecutor::create(1);
  auto eef = EEF<1>::create<EEFType::G2, MotorType::DM>();

  std::string can = program.get<std::string>("-i");
  std::string UDP_SERVER_IP = program.get<std::string>("-r");
  int UDP_SERVER_PORT = program.get<int>("-p");
  double eff = program.get<double>("-f");

  if (!eef->init(executor->get_io_context(), can, 250_hz)) {
    std::cerr << "Failed to initialize eef" << std::endl;
    return 1;
  }

  // Setup UDP socket
  boost::asio::io_context io_context;
  boost::asio::ip::udp::socket socket(io_context);
  boost::asio::ip::udp::endpoint server_endpoint(boost::asio::ip::make_address(UDP_SERVER_IP), UDP_SERVER_PORT);
  try {
    socket.open(boost::asio::ip::udp::v4());
    std::cout << "UDP socket opened, sending data to " << UDP_SERVER_IP << ":" << UDP_SERVER_PORT << std::endl;
  } catch (std::exception& e) {
    std::cerr << "Failed to open UDP socket: " << e.what() << std::endl;
    eef->disable();
    eef->uninit();
    return 1;
  }

  auto t = std::thread([&]() {
    while (running) {
      std::this_thread::sleep_for(std::chrono::milliseconds(4));
      if (eef->state().is_valid) {
        auto state = eef->state();
        std::vector<double> pos(state.pos.begin(), state.pos.end());
        std::vector<double> vel(state.vel.begin(), state.vel.end());
        std::vector<double> eff(state.eff.begin(), state.eff.end());
        std::string state_string = boost::json::serialize(commandToJson("state", pos, vel, eff));

        try {
          socket.send_to(boost::asio::buffer(state_string), server_endpoint);
        } catch (std::exception& e) {
          std::cerr << "Failed to send state data: " << e.what() << std::endl;
        }
      }
    }
  });

  eef->enable();
  eef->set_param("control_mode", static_cast<uint32_t>(MotorControlMode::MIT));

  while (running) {
    eef->mit({0.0, 0.0, eff, 0.0, 0.0});
    std::this_thread::sleep_for(4ms);
  }

  t.join();
  socket.close();
  eef->disable();
  eef->uninit();
  return 0;
}

File: example/python/product/g2_force.py
import argparse
import airbot_hardware_py as ah
import time
import threading
import json
import socket
import math

# UDP server configuration (PlotJuggler's listening address)
UDP_SERVER_IP = "127.0.0.1"
UDP_SERVER_PORT = 9870

# Create a UDP socket
sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
server_address = (UDP_SERVER_IP, UDP_SERVER_PORT)

print(f"Sending data to {UDP_SERVER_IP}:{UDP_SERVER_PORT}")


def send_to_plotjuggler(eef, stop_event):
    while not stop_event.is_set():
        current_time = time.time()
        motor_num = 1
        state_forces = [0.0] * motor_num

        for i in range(motor_num):
            if eef.state().is_valid:
                state_forces[i] = eef.state().eff[i]

        # Create the JSON object
        json_obj = {
            "state/force": state_forces,
            "timestamp": current_time,
        }

        # Serialize the JSON object to a string and then encode to bytes
        message = json.dumps(json_obj).encode("utf-8")

        try:
            sock.sendto(message, server_address)
        except Exception as e:
            print(f"Failed to send state data: {e}")

        # Wait before the next round of prints
        time.sleep(0.01)


def main():
    # 设置命令行参数解析
    parser = argparse.ArgumentParser(
        description="单独 G2 夹爪力控制 - 在不同的位置施加恒定的夹取力"
    )
    parser.add_argument(
        "-f", "--force", type=float, required=True, help="夹爪施加的力，例如 5"
    )
    parser.add_argument(
        "-i", "--can", default="can0", help="can口名称 (如: can0, can1)"
    )

    args = parser.parse_args()

    # 获取解析后的参数值
    force = args.force
    can_interface = args.can

    executor = ah.create_asio_executor(1)
    io_context = executor.get_io_context()

    eef = ah.EEF1.create(ah.EEFType.G2, ah.MotorType.DM)
    if not eef.init(io_context, can_interface, 250):
        raise RuntimeError("The EEF initialization failed.")
    eef.enable()
    eef.set_param("control_mode", ah.MotorControlMode.MIT)
    stop_event = threading.Event()
    print_thread = threading.Thread(target=send_to_plotjuggler, args=(eef, stop_event))
    print_thread.daemon = True
    print_thread.start()
    try:
        while True:
            cmd = ah.EEFCommand1()
            cmd.eff = [force]
            eef.mit(cmd)
            time.sleep(0.004)
    except KeyboardInterrupt:
        print("\n检测到Ctrl+C，正在停止机械臂并释放资源...")
    finally:
        stop_event.set()
        if print_thread.is_alive():
            print_thread.join()

        eef.disable()
        eef.uninit()


if __name__ == "__main__":
    main()
